python/ngen_conf/src/ngen/config/path_pair/path_pair.py

```diff
@@ -0,0 +1,277 @@
+    writing, and de/serializing to and from a the collection of `T`'s. This type enables treating a
+    collection of paths, that follow some naming convention, as a single path object. Similarly to
+    `pathlib.Path`, creating a `PathPairCollection` will return a OS specific subclass
+    (`PosixPathPairCollection` or `WindowsPathPairCollection`).
+
+    The default reader and writer implementations operate on disk. Alternative reader and writer
+    types can be provided assuming they following the `ngen.config.path_path.protocol` `Reader` and
+    `Writer` protocols (interfaces). This could, for example, enable reading and writing from the
+    network.
+
+    serializer's and deserializer's are type specific, and thus have no default. However,
+    `ngen.config.path_path.common` contains generic serializers and deserializer that may be of
+    interest (e.g. `pydantic_serializer` and `pydantic_deserializer`). Alternative serializer and
+    deserializer types can be provided assuming they follow the `ngen.config.path_path.protocol`
+    `Serializer` and `Deserializer` protocols (interfaces).
+    """
+
+    def __new__(
+        cls,
+        *args: StrPath,
+        pattern: str,
+        inner: Optional[List["PosixPathPair[T]"]] = None,
+        reader: Reader = path_reader,
+        writer: Writer = path_writer,
+        serializer: Optional[Serializer[T]] = None,
+        deserializer: Optional[Deserializer[T]] = None,
+        **kwargs: Any,
+    ) -> Union["WindowsPathPairCollection[T]", "PosixPathPairCollection[T]"]:
+        cls = (
+            WindowsPathPairCollection[T]
+            if os.name == "nt"
+            else PosixPathPairCollection[T]
+        )
+
+        template_str = Path(*args)
+
+        if inner is None:
+            inner = []
+
+        prefix, _, suffix = template_str.name.partition(pattern)
+        for item in inner:
+            if PathPairCollection._get_id(item, prefix, suffix) == "":
+                raise ValueError(
+                    f"Filename not derived from template and pattern, {template_str.name!r} {pattern!r}: {item}"
+                )
+
+        self: Union[
+            WindowsPathPairCollection[T], PosixPathPairCollection[T]
+        ] = Path.__new__(
+            cls,
+            *args,
+            inner=inner,
+            reader=reader,
+            writer=writer,
+            serializer=serializer,
+            deserializer=deserializer,
+            **kwargs,
+        )
+        self._inner = inner
+        self._serializer = serializer
+        self._deserializer = deserializer
+        self._reader = reader
+        self._writer = writer
+        self._pattern = pattern
+        return self
+
+    @staticmethod
+    def _get_id(
+        p: "PosixPathPair[T]",
+        prefix: str,
+        suffix: str,
+    ) -> str:
+        assert prefix != "" and suffix != ""
+
+        return p.name[len(prefix) : -len(suffix)]
+
+    @classmethod
```

Comment:
what is the intent of having these  wrappers to existing Path class methods?

Reply:
IMO it didnt make sense for `cwd` to return an empty `PathPairCollection`. Instead, I thought it made more sense to just return a `Path` instance.

Reply:
I guess I'm still not sure why/what you need `PathPairCollection.cwd()` for vs just using `Path.cwd()`. Why support such a weak wrapper?

Reply:
Ok, understanding now that this is just an extreme edge case protection, on the off chance someone DID decide to call `PathPairCollection.cwd(), they would get back a `Path.cwd()` instead of a `PathPairCollection( os.get_cwd )` which would then require additional args (e.g. `pattern` to construct, and without a reasonable default arg for `pattern` that is tricky...